File: src/algorithms/python/topo-sort.py
import json

def emit(event_type, **kwargs):
    event = {"type": event_type}
    event.update(kwargs)
    print(json.dumps(event))

from collections import deque

def topo_sort(graph):
    n = len(graph)
    in_degree = [0] * n
    
    # Calculate in-degrees
    for u in range(n):
        for v, w, edge_id in graph[u]:
            in_degree[v] += 1
            
    queue = deque()
    for i in range(n):
        if in_degree[i] == 0:
            queue.append(i)
            emit("GRAPH_NODE_HIGHLIGHT", nodeId=f"n{i}", status="queued")
            
    result = []
    
    while queue:
        u = queue.popleft()
        emit("GRAPH_NODE_HIGHLIGHT", nodeId=f"n{u}", status="current")
        result.append(u)
        
        for v, w, edge_id in graph[u]:
            emit("GRAPH_EDGE_HIGHLIGHT", edgeId=edge_id, accepted=True)
            in_degree[v] -= 1
            if in_degree[v] == 0:
                queue.append(v)
                emit("GRAPH_NODE_HIGHLIGHT", nodeId=f"n{v}", status="queued")
                
        emit("GRAPH_NODE_HIGHLIGHT", nodeId=f"n{u}", status="removed")
        
    if len(result) != n:
        emit("SYSTEM_LOG", message="Graph has a cycle! Topological sort not possible.", level="WARN")
    else:
        emit("SYSTEM_LOG", message=f"Topological Sort order: {result}", level="INFO")

if __name__ == "__main__":
    graph = [
        [(1, 1, "e0"), (2, 1, "e1")],
        [(3, 1, "e2")],
        [(3, 1, "e3")],
        []
    ]
    topo_sort(graph)

File: src/algorithms/cpp/topo-sort.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <algorithm>

using namespace std;

struct Edge { int node; string id; };

void emitNodeHighlight(int node) {
    cout << "{\"type\":\"GRAPH_NODE_HIGHLIGHT\",\"nodeId\":\"n" << node << "\"}" << endl;
}
void emitEdgeHighlight(string edgeId, bool accepted) {
    cout << "{\"type\":\"GRAPH_EDGE_HIGHLIGHT\",\"edgeId\":\"" << edgeId << "\",\"accepted\":" << (accepted ? "true" : "false") << "}" << endl;
}
void emitSystemLog(string message) {
    cout << "{\"type\":\"SYSTEM_LOG\",\"message\":\"" << message << "\",\"level\":\"INFO\"}" << endl;
}

void topoVisit(int u, vector<vector<Edge>>& graph, vector<bool>& visited, vector<int>& order) {
    visited[u] = true;
    emitNodeHighlight(u);
    for (auto& edge : graph[u]) {
        if (!visited[edge.node]) {
            emitEdgeHighlight(edge.id, true);
            topoVisit(edge.node, graph, visited, order);
        }
    }
    order.push_back(u);
}

void emitInit(const vector<vector<Edge>>& graph) {
    cout << "{\"type\":\"INIT\",\"graph\":{\"nodes\":[";
    for (int i = 0; i < (int)graph.size(); i++) {
        if (i > 0) cout << ",";
        cout << "{\"id\":\"n" << i << "\",\"label\":\"" << i << "\",\"x\":" << (i%3)*120-120 << ",\"y\":" << (i/3)*120-60 << "}";
    }
    cout << "],\"edges\":[";
    bool first = true;
    for (int u = 0; u < (int)graph.size(); u++) {
        for (const auto& e : graph[u]) {
            if (!first) cout << ",";
            first = false;
            cout << "{\"id\":\"" << e.id << "\",\"from\":\"n" << u << "\",\"to\":\"n" << e.node << "\"}";
        }
    }
    cout << "],\"isDirected\":true}}" << endl;
}

int main() {
    vector<vector<Edge>> graph(5);
    graph[0] = {{1, "e0"}, {2, "e1"}};
    graph[1] = {{3, "e2"}};
    graph[2] = {{3, "e3"}, {4, "e4"}};
    graph[3] = {{4, "e5"}};
    graph[4] = {};

    emitInit(graph);
    vector<bool> visited(5, false);
    vector<int> order;
    for (int i = 0; i < 5; i++) {
        if (!visited[i]) {
            topoVisit(i, graph, visited, order);
        }
    }
    reverse(order.begin(), order.end());
    
    string res = "Topological Order: ";
    for (int i = 0; i < (int)order.size(); i++) {
        if (i > 0) res += " -> ";
        res += "n" + to_string(order[i]);
    }
    emitSystemLog(res);
    return 0;
}
